### ISSUE TITLE
Cannot run ansible modules on VMware ESXi via SSH

### ISSUE THREAD

**Issue by enkiusz** [ansible/ansible#25178]
##### ISSUE TYPE
 - Bug Report

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.4.0
  config file = 
  configured module search path = [u'/home/enki/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible
  executable location = /home/enki/devel/ansible-esxi-fix/bin/ansible
  python version = 2.7.12 (default, Dec 16 2016, 00:14:46) [GCC 4.9.3]
```

##### CONFIGURATION
Default.

##### OS / ENVIRONMENT
Gentoo system managing a VMware ESXi machine.

##### SUMMARY
Can't run any modules via SSH on the ESXi machine, ansible module doesn't like the six module in /lib64/python3.5/site-packages on the target esxi host. After moving six.pyc in the above directory aside (or removing it entirely) the ansible modules begin to work as they are now including their internal six.py from the zip.

I suspect that the reason for the wrong behaviour is that the ansible module code expects a six.py ("source") file but only finds six.pyc which it tries to load as a source file triggering an exception.

##### STEPS TO REPRODUCE
1. Take a free ESXi image and install it on a server.
2. Enable SSH using the admin UI.
3. Setup SSH key authentication for your management machine.
4. Try to run any ansible module via SSH.
5. Setup the following virtualenv and try to run ansible:

```
$ virtualenv -p python2.7 ansible-esxi-fix
$ cd ansible-esxi-fix/
$ source bin/activate
$ git clone https://github.com/ansible/ansible.git
$ cd ansible/
$ python setup.py build
$ python setup.py install
$ echo "newton ansible_host=192.168.4.110 ansible_user=root" > inventory
$ ssh root@192.168.4.110 ls -l /etc/vmware | head -n 3
kex protocol error: type 7 seq 12
total 748
-r--r--r--    1 root     root          3041 Jan  6 02:51 BootbankFunctions.sh
drwxr-xr-x    1 root     root           512 Jan  6 03:12 autodeploy
```

##### EXPECTED RESULTS
After removing the six.pyc file ansible works correctly:

[root@newton:/etc/vmware] rm /lib64/python3.5/site-packages/six.pyc 

```
$ ansible -i inventory all -m setup
newton | SUCCESS => {
    "ansible_facts": {
        "ansible_apparmor": {
            "status": "disabled"
        }, 
        "ansible_architecture": "x86_64", 
        "ansible_date_time": {
            "date": "2017-05-30", 
            "day": "30", 
            "epoch": "1496184937", 
            "hour": "22", 
            "iso8601": "2017-05-30T22:55:37Z", 
            "iso8601_basic": "20170530T225537227354", 
            "iso8601_basic_short": "20170530T225537", 
            "iso8601_micro": "2017-05-30T22:55:37.227651Z", 
            "minute": "55", 
            "month": "05", 
            "second": "37", 
            "time": "22:55:37", 
            "tz": "UTC", 
            "tz_offset": "+0000", 
            "weekday": "Tuesday", 
            "weekday_number": "2", 
            "weeknumber": "22", 
            "year": "2017"
        }, 
        "ansible_distribution": "VMkernel", 
        "ansible_distribution_release": "6.5.0", 
        "ansible_distribution_version": "#1 SMP Release build-4887370 Jan  5 2017 19:17:59", 
        "ansible_dns": {
            "nameservers": [
                "192.168.4.2"
            ]
        }, 
[...]
    "changed": false
}
```

##### ACTUAL RESULTS
```
$ ansible -vvvv -i inventory all -m setup
ansible 2.4.0
  config file = 
  configured module search path = [u'/home/enki/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible
  executable location = /home/enki/devel/ansible-esxi-fix/bin/ansible
  python version = 2.7.12 (default, Dec 16 2016, 00:14:46) [GCC 4.9.3]
No config file found; using defaults
setting up inventory plugins
Parsed /home/enki/devel/ansible-esxi-fix/ansible/inventory inventory source with ini plugin
Loading callback plugin minimal of type stdout, v2.0 from /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible/plugins/callback/__init__.pyc
META: ran handlers
Using module_utils file /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible/module_utils/basic.py
Using module_utils file /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible/module_utils/facts.py
Using module_utils file /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible/module_utils/_text.py
Using module_utils file /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible/module_utils/pycompat24.py
Using module_utils file /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible/module_utils/six/__init__.py
Using module_utils file /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible/module_utils/six/_six.py
Using module file /home/enki/devel/ansible-esxi-fix/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible/modules/system/setup.py
<192.168.4.110> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.4.110> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 -o ControlPath=/home/enki/.ansible/cp/fce537cc11 192.168.4.110 '/bin/sh -c '"'"'echo ~ && sleep 0'"'"''
<192.168.4.110> (0, '/\n', 'OpenSSH_7.3p1-hpn14v11, OpenSSL 1.0.2k  26 Jan 2017\r\ndebug1: Reading configuration data /home/enki/.ssh/config\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 6187\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<192.168.4.110> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.4.110> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 -o ControlPath=/home/enki/.ansible/cp/fce537cc11 192.168.4.110 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /.ansible/tmp/ansible-tmp-1496181193.31-18628934631670 `" && echo ansible-tmp-1496181193.31-18628934631670="` echo /.ansible/tmp/ansible-tmp-1496181193.31-18628934631670 `" ) && sleep 0'"'"''
<192.168.4.110> (0, 'ansible-tmp-1496181193.31-18628934631670=/.ansible/tmp/ansible-tmp-1496181193.31-18628934631670\n', 'OpenSSH_7.3p1-hpn14v11, OpenSSL 1.0.2k  26 Jan 2017\r\ndebug1: Reading configuration data /home/enki/.ssh/config\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 6187\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<192.168.4.110> PUT /tmp/tmpSsfj1X TO /.ansible/tmp/ansible-tmp-1496181193.31-18628934631670/setup.py
<192.168.4.110> SSH: EXEC sftp -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 -o ControlPath=/home/enki/.ansible/cp/fce537cc11 '[192.168.4.110]'
<192.168.4.110> (0, 'sftp> put /tmp/tmpSsfj1X /.ansible/tmp/ansible-tmp-1496181193.31-18628934631670/setup.py\n', 'OpenSSH_7.3p1-hpn14v11, OpenSSL 1.0.2k  26 Jan 2017\r\ndebug1: Reading configuration data /home/enki/.ssh/config\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 6187\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug2: Remote version: 3\r\ndebug2: Server supports extension "posix-rename@openssh.com" revision 1\r\ndebug2: Server supports extension "statvfs@openssh.com" revision 2\r\ndebug2: Server supports extension "fstatvfs@openssh.com" revision 2\r\ndebug2: Server supports extension "hardlink@openssh.com" revision 1\r\ndebug2: Server supports extension "fsync@openssh.com" revision 1\r\ndebug3: Sent message fd 6 T:16 I:1\r\ndebug3: SSH_FXP_REALPATH . -> / size 0\r\ndebug3: Looking up /tmp/tmpSsfj1X\r\ndebug3: Sent message fd 6 T:17 I:2\r\ndebug3: Received stat reply T:101 I:2\r\ndebug1: Couldn\'t stat remote file: No such file or directory\r\ndebug3: Sent message SSH2_FXP_OPEN I:3 P:/.ansible/tmp/ansible-tmp-1496181193.31-18628934631670/setup.py\r\ndebug3: Sent message SSH2_FXP_WRITE I:4 O:0 S:32768\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 4 32768 bytes at 0\r\ndebug3: Sent message SSH2_FXP_WRITE I:5 O:32768 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:6 O:65536 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:7 O:98304 S:12844\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 5 32768 bytes at 32768\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 6 32768 bytes at 65536\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 7 12844 bytes at 98304\r\ndebug3: Sent message SSH2_FXP_CLOSE I:4\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<192.168.4.110> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.4.110> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 -o ControlPath=/home/enki/.ansible/cp/fce537cc11 192.168.4.110 '/bin/sh -c '"'"'chmod u+x /.ansible/tmp/ansible-tmp-1496181193.31-18628934631670/ /.ansible/tmp/ansible-tmp-1496181193.31-18628934631670/setup.py && sleep 0'"'"''
<192.168.4.110> (0, '', 'OpenSSH_7.3p1-hpn14v11, OpenSSL 1.0.2k  26 Jan 2017\r\ndebug1: Reading configuration data /home/enki/.ssh/config\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 6187\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<192.168.4.110> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.4.110> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 -o ControlPath=/home/enki/.ansible/cp/fce537cc11 -tt 192.168.4.110 '/bin/sh -c '"'"'/usr/bin/python /.ansible/tmp/ansible-tmp-1496181193.31-18628934631670/setup.py; rm -rf "/.ansible/tmp/ansible-tmp-1496181193.31-18628934631670/" > /dev/null 2>&1 && sleep 0'"'"''
<192.168.4.110> (0, 'Traceback (most recent call last):\r\n  File "/tmp/ansible_tx0y2tyu/ansible_module_setup.py", line 135, in <module>\r\n    from ansible.module_utils.basic import *\r\n  File "<frozen importlib._bootstrap>", line 969, in _find_and_load\r\n  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked\r\n  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked\r\n  File "<frozen importlib._bootstrap>", line 634, in _load_backward_compatible\r\n  File "/tmp/ansible_tx0y2tyu/ansible_modlib.zip/ansible/module_utils/basic.py", line 164, in <module>\r\n  File "<frozen importlib._bootstrap>", line 969, in _find_and_load\r\n  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked\r\n  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked\r\n  File "<frozen importlib._bootstrap>", line 634, in _load_backward_compatible\r\n  File "/tmp/ansible_tx0y2tyu/ansible_modlib.zip/ansible/module_utils/six/__init__.py", line 116, in <module>\r\nValueError: source code string cannot contain null bytes\r\n', 'OpenSSH_7.3p1-hpn14v11, OpenSSL 1.0.2k  26 Jan 2017\r\ndebug1: Reading configuration data /home/enki/.ssh/config\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 6187\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to 192.168.4.110 closed.\r\n')
newton | FAILED! => {
    "changed": false, 
    "failed": true, 
    "module_stderr": "OpenSSH_7.3p1-hpn14v11, OpenSSL 1.0.2k  26 Jan 2017\r\ndebug1: Reading configuration data /home/enki/.ssh/config\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 6187\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to 192.168.4.110 closed.\r\n", 
    "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_tx0y2tyu/ansible_module_setup.py\", line 135, in <module>\r\n    from ansible.module_utils.basic import *\r\n  File \"<frozen importlib._bootstrap>\", line 969, in _find_and_load\r\n  File \"<frozen importlib._bootstrap>\", line 958, in _find_and_load_unlocked\r\n  File \"<frozen importlib._bootstrap>\", line 664, in _load_unlocked\r\n  File \"<frozen importlib._bootstrap>\", line 634, in _load_backward_compatible\r\n  File \"/tmp/ansible_tx0y2tyu/ansible_modlib.zip/ansible/module_utils/basic.py\", line 164, in <module>\r\n  File \"<frozen importlib._bootstrap>\", line 969, in _find_and_load\r\n  File \"<frozen importlib._bootstrap>\", line 958, in _find_and_load_unlocked\r\n  File \"<frozen importlib._bootstrap>\", line 664, in _load_unlocked\r\n  File \"<frozen importlib._bootstrap>\", line 634, in _load_backward_compatible\r\n  File \"/tmp/ansible_tx0y2tyu/ansible_modlib.zip/ansible/module_utils/six/__init__.py\", line 116, in <module>\r\nValueError: source code string cannot contain null bytes\r\n", 
    "msg": "MODULE FAILURE", 
    "rc": 0
}
```

**Comment by abadger** [comment]
I'm working on a PR for 2.4.0 that includes (among a great number of other things) always using the bundled six library: https://github.com/ansible/ansible/pull/27254/commits/8b9ecfd56f9d8b276f0bc8c2e607674f2a1be7a2

I believe that will fix this issue (there's one other code path that could be causing the wrong six to be used but I believe it's the module_utils code which is doing this and not the module_common code.)  If all goes well, that change will be merged later today.

Note that Operating System vendors may decide to use some variant of the previous code to unbundle the six library that we ship.  Until we ship it and see how they react we aren't going to know whether they do or not and whether their implementations suffer from this same defect or not.

**Comment by abadger** [comment]
References #27254

**Comment by abadger** [comment]
#27254 has been merged to devel.  If you could test the devel branch I believe it will fix this issue for you.

needs_info

**Comment by abadger** [comment]
Closing ticket as it should be fixed by #27254

**Comment by enkiusz** [comment]
It's fixed, thanks. Sorry for the late reply.
